### PR TITLE
Added MulByPow2 and MulByFloorPow2 ops

### PR DESCRIPTION
**Pull request by johnplatts**
Resolves issue #2266.

Should `MulByFloorPow2(a, b)` have implementation-defined behavior in the case where `a[i]` is NaN and `b[i]` is negative infinity as the AVX3 _mm_scalef_ph/ps/pd ops return zero in the case where `a[i]` is NaN and `b[i]` is negative infinity whereas the default `MulByFloorPow2(a, b)` implementation returns NaN in any lanes where `a[i]` is NaN and `b[i]` is negative infinity?

### CODE REVIEW

File: hwy/ops/generic_ops-inl.h [hunk 1]
```diff
@@ -970,6 +970,154 @@ HWY_API VFromD<RebindToSigned<DFromV<V>>> FloorInt(V v) {
...
+
+  // If kZeroOutSignUsingSatSub is true, then val_for_exp_sub will be bitcasted
+  // to a vector that has a smaller lane size than TU for the SaturatedSub
+  // operation below.
+  constexpr bool kZeroOutSignUsingSatSub =
+      ((sizeof(TExpSatSub) * 8) == static_cast<size_t>(kNumOfExpBits));
+
+  // If kZeroOutSignUsingSatSub is true, then the upper
+  // (sizeof(TU) - sizeof(TExpSatSub)) * 8 bits of kExpDecrBy1Bits will be all
+  // ones and the lower sizeof(TExpSatSub) * 8 bits of kExpDecrBy1Bits will be
```
Comment (jan-wassenberg) [inline hwy/ops/generic_ops-inl.h:1043]: This is a bit confusing, we're saying the lower bits are equal to 0...01. The comment is accurate but possible to misunderstand. Not sure there is a better way to say it, though.

**Review by jan-wassenberg — APPROVED** [review]
I agree impl-defined for the NaN * -inf case is reasonable.

Very nice, a combination of wizard-like generic/fallback implementation, plus thorough testing and comment. Thanks for adding!

File: hwy/ops/generic_ops-inl.h (hunk 1 above)
Comment (johnplatts) [reply hwy/ops/generic_ops-inl.h:1043]: What I meant to say is that if `ExponentBits<TF>()` is a power of 2 that is greater than or equal to 8 (which is true if `TF` is `float` as `ExponentBits<float>()` is equal to 8), then the upper `sizeof(TF)*8 - ExponentBits<TF>()` bits can be zeroed out using a SaturatedSub op using a bitcasted vector that has a lane type of `UnsignedFromSize<(ExponentBits<TF>() / 8)>` on targets other than HWY_SCALAR.

In other words, the following U8 SaturatedSub op can simultaneously decrement the lower 8 bits of an U32 lane by 1 (with unsigned saturation) and zero out the upper 24 bits of each U32 lane on targets other than HWY_SCALAR:
`BitCast(du32, SaturatedSub(BitCast(du8, v), BitCast(du8, Set(du32, uint32_t{0xFFFFFF01u}))))`

Comment (jan-wassenberg) [reply hwy/ops/generic_ops-inl.h:1043]: Got it, thanks. SaturatedSub is surprisingly useful :)

**Comment by johnplatts** [comment]
> Internal MSVC builds are failing due to timeout. Can we shorten/simplify the test for MSVC?

I have reimplemented TestMulByPow2 and TestMulByFloorPow2, which significantly speeds up build times with MSVC.

I have also discovered that `MulByFloorPow2(nan, inf)` is implementation-defined as AVX3 `_mm_scalef_*(nan, +inf)` returns positive infinity whereas the default implementation of `MulByFloorPow2(nan, inf)` returns NaN.

I have updated TestMulByFloorPow2 to check that `MulByFloorPow2(nan, -inf)` returns 0 or NaN and that `MulByFloorPow2(nan, +inf)` returns positive infinity or NaN.

**Comment by jan-wassenberg** [comment]
Thanks for refactoring the test. CI passed now :)

Interesting behavior from the AVX3 instructions, oh well. I agree it's fine to have differences for nan and +/- inf.
